### PR TITLE
Feat: Rework multimodal support (mllama, llava, pixtral, qwen2, qwen25, gemma3, mistral3)

### PR DESCRIPTION
**Pull request by NanoCode012**
# Description

This PR continues the work in #2170 and adds multimodal support for:
- mllama
- llava15
- pixtral
- qwen2
- qwen25
- gemma3
- mistral3

We reworked the data collator to use `apply_chat_template` to tokenize via the new transformers 4.49.0 update.

Requires `transformers==4.50.0` released today.

## How has this been tested?

The example configs are ran on cloud pod (but with lower settings to save vram) and verified both train/eval loss decrease appropriately.

We will run against external benchmarks as well.

### CODE REVIEW

File: docs/config.qmd [hunk 1]
```diff
@@ -578,6 +578,14 @@ resume_from_checkpoint:
 # Be careful with this being turned on between different models.
 auto_resume_from_checkpoints: false
 
+## Multimodal section
+# int | tuple[int, int] | None . Size to resize images to, width x height.
```
Comment (djsaunde) [inline docs/config.qmd:590]: I like the typing indication here in the config docs. I wonder if there's a better way to do this, e.g.,

```
image_size: int | tuple[int, int] | None = None
```

This would just be duplicating the typing indication in the Axolotl config pydantic model(s).

Maybe even better would be to just generate these docs from the Pydantic models themselves. We could use `Field`s with `json_schema_extra` populated with the comments that we have here.

Just an idea. It might help reduce duplication of efforts.

File: docs/multimodal.qmd
Comment (djsaunde) [inline docs/multimodal.qmd:1]: Add to `_quarto.yml`?

File: docs/multimodal.qmd [hunk 2]
```diff
@@ -1,28 +1,171 @@
...
+
+Here are the hyperparams you'll need to use to finetune a multimodal model.
 
 ```yaml
-base_model: alpindale/Llama-3.2-11B-Vision-Instruct
 processor_type: AutoProcessor
+
 skip_prepare_dataset: true
+remove_unused_columns: false
+sample_packing: false
```
Comment (djsaunde) [inline docs/multimodal.qmd]: I would write a quick comment as to why these settings are needed.

**Comment by ehartford** [comment]
thanks for this; I'm currently stuck using llama factory for this feature.

File: src/axolotl/processing_strategies/__init__.py
Comment (djsaunde) [inline src/axolotl/processing_strategies/__init__.py:1]: Nit: don't put meaningful code in `__init__.py` files, they're meant for module initialization.

If you make this change, and you want e.g. `ProcessingStrategy` to be import-able from `axolotl.processing_strategies`, import it in `__init__.py`.

Better yet, just define `axolotl/utils/processing_strategies.py` and remove the folder since this is just a single file.

File: src/axolotl/processing_strategies/__init__.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+        Converts different conversation formats to OpenAI format with 'messages'.
+        Supports two formats:
+        1. OpenAI format with 'messages'
+        2. Legacy format with 'conversations'
+
+        Args:
+            examples: list of conversation dictionaries
+
+        Returns:
+            dict in OpenAI format with 'messages' key
```
Comment (djsaunde) [inline src/axolotl/processing_strategies/__init__.py]: nit: list of dicts

File: src/axolotl/processing_strategies/__init__.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+            result = deepcopy(example)
+            result.pop("conversations")
+            result["messages"] = messages
+            return result
+
+        def convert_messages_to_multimedia_messages(messages: list[dict]) -> list[dict]:
+            """Convert regular messages format to Messages format with content type"""
+
+            new_messages = []
+            for convo in messages:
```
Comment (djsaunde) [inline src/axolotl/processing_strategies/__init__.py]: nit:

```suggestion
            for message in messages:
```

File: src/axolotl/utils/collators/mm_chat.py [hunk 5]
```diff
@@ -2,15 +2,16 @@
...
 from typing import Any, Optional, Union
 
-from PIL import Image
-from transformers import PreTrainedTokenizerBase, ProcessorMixin
+import torch
+from transformers import PreTrainedTokenizerBase
 from transformers.data.data_collator import DataCollatorMixin
 from transformers.utils import PaddingStrategy
 
+from ...processing_strategies import ProcessingStrategy
```
Comment (djsaunde) [inline src/axolotl/utils/collators/mm_chat.py]: nit: absolute imports are preferred

File: src/axolotl/utils/collators/mm_chat.py [hunk 6]
```diff
@@ -19,174 +20,76 @@ class MultiModalChatDataCollator(DataCollatorMixin):
...
-        self, examples: list[Union[list[int], Any, dict[str, Any]]]
-    ) -> dict[str, Any]:
-        # Handle dict or lists with proper padding and conversion to tensor.
+    def torch_call(self, examples: list[dict]) -> dict[str, Any]:
+        return self.process_rows(examples)
 
-        return self.__class__.process_rows(
-            examples, self.processor, self.chat_template, self.max_images
-        )
+    def process_rows(
```
Comment (djsaunde) [inline src/axolotl/utils/collators/mm_chat.py:37]: nit: missing return type

File: src/axolotl/utils/collators/mm_chat.py [hunk 7]
```diff
@@ -19,174 +20,76 @@ class MultiModalChatDataCollator(DataCollatorMixin):
...
+        self,
+        examples: list[dict],
+    ):
+        # Preprocess the examples
+        examples = self.processing_strategy.preprocess(examples)
```
Comment (djsaunde) [inline src/axolotl/utils/collators/mm_chat.py]: should we consider making the ProcessingStrategy class callable (implement `__call__` in place of`preprocess`) ?

File: src/axolotl/utils/collators/mm_chat.py [hunk 8]
```diff
@@ -19,174 +20,76 @@ class MultiModalChatDataCollator(DataCollatorMixin):
...
-
-            return images
+            # TODO: Check if need handling for len(input_ids) > sequence_len
 
-        # Preprocess the examples
-        examples = _preprocess(examples)
+            # Add the processed tensors to our batch
+            for key in result.keys():
+                if key not in batch:
+                    batch[key] = []
```
Comment (djsaunde) [inline src/axolotl/utils/collators/mm_chat.py:66]: If we're making this check, I think we can init the batch dict as `{}`.

File: src/axolotl/utils/config/__init__.py [hunk 9]
```diff
@@ -158,7 +158,8 @@ def normalize_config(cfg):
 
     cfg.is_multimodal = (
         hasattr(model_config, "model_type")
-        and model_config.model_type in ["llava", "mllama"]
+        and model_config.model_type
+        in ["llava", "mllama", "qwen2_vl", "qwen2_5_vl", "gemma3", "mistral3"]
```
Comment (djsaunde) [inline src/axolotl/utils/config/__init__.py]: Nit: constantize this.

File: docs/config.qmd (hunk 1 above)
Comment (NanoCode012) [reply docs/config.qmd:590]: Yep, I agree, that's why I'm really excited with the quartodocs :)

File: docs/multimodal.qmd
Comment (NanoCode012) [reply docs/multimodal.qmd:1]: I was just updating existing file

File: docs/config.qmd (hunk 1 above)
Comment (djsaunde) [reply docs/config.qmd:590]: I'm not sure quartodoc can do it, but it should be relatively easy to do with the script if not!

File: src/axolotl/utils/models.py [hunk 10]
```diff
@@ -70,8 +74,12 @@
 LOG = logging.getLogger(__name__)
 
 MULTIMODEL_AUTO_MODEL_MAPPING = {
```
Comment (djsaunde) [inline src/axolotl/utils/models.py]: Just realized there's a typo here.

```suggestion
MULTIMODAL_AUTO_MODEL_MAPPING = {
```

**Comment by winglian** [comment]
should probably merge/rebase from main to get the fixes from #2436 to fix the ring attention test

**Review by winglian — APPROVED** [review]
